### ISSUE TITLE
tf.keras.metrics.Metric not returning actual result value over batches

### ISSUE THREAD

**Issue by mdalvi** [tensorflow/tensorflow#47099]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Microsoft Windows 10 Enterprise Version 10.0.18363 Build 18363
- TensorFlow installed from (source or binary): binary (pip install tensorflow)
- TensorFlow version : v2.4.0-49-g85c8b2a817f 2.4.1
- Python version: 3.7.6

**Describe the current behavior**
The `result()` method of the `tf.keras.metrics.Metric` class as defined [here](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Metric#result) is supposed to return the calculation over state variables. The values it returns seem to be affected by some kind of decorator over it that is not allowing to return of the actual value rather a generalized version of value considering the batch number

**Describe the expected behavior**
It should return what it is calculating without any transformations 

**Standalone code to reproduce the issue**
```
class AddAllOnes(tf.keras.metrics.Metric):
  """ A simple metric that adds all the one's in current batch and suppose to return the total ones seen at every end of batch"""
    def __init__(self, name="add_all_ones", **kwargs):
        super(AddAllOnes, self).__init__(name=name, **kwargs)
        self.total = self.add_weight(name="total", initializer="zeros")

    def update_state(self, y_true, y_pred, sample_weight=None):    
        self.total.assign_add(tf.cast(tf.reduce_sum(y_true), dtype=tf.float32))
        
    def result(self):
        print('')
        print('inside result...', self.total)
        return self.total

X_train = np.random.random((512, 8))
y_train = np.random.randint(0, 2, (512, 1))

K.clear_session()
model_inputs = Input(shape=(8,))
model_unit = Dense(256, activation='linear', use_bias=False)(model_inputs)
model_unit = BatchNormalization()(model_unit)
model_unit = Activation('sigmoid')(model_unit)
model_outputs = Dense(1, activation='sigmoid')(model_unit)
optim = Adam(learning_rate=0.001)
model = Model(inputs=model_inputs, outputs=model_outputs)
model.compile(loss='binary_crossentropy', optimizer=optim, metrics=[AddAllOnes()], run_eagerly=True)
model.fit(X_train, y_train, verbose=1, batch_size=32)
```

**Other info / logs**
Do you see the difference between the printed value within the result and the returned value on the epoch logs (except the first iteration everything is different)

E.g. In the first iteration the metric saw a total of 19 ones and so returned 19.0 in epoch logs however in the 5th iteration it is supposed to return 79 but returned 49 in epoch logs! 

```
inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=19.0>
 1/16 [>.............................] - ETA: 0s - loss: 0.6744 - add_all_ones: 19.0000
inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=35.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=53.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=61.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=79.0>
 5/16 [========>.....................] - ETA: 0s - loss: 0.7072 - add_all_ones: 49.4000
inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=93.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=108.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=120.0>
 8/16 [==============>...............] - ETA: 0s - loss: 0.7084 - add_all_ones: 71.0000
inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=135.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=149.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=158.0>
11/16 [===================>..........] - ETA: 0s - loss: 0.7074 - add_all_ones: 91.8182
inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=169.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=183.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=203.0>
14/16 [=========================>....] - ETA: 0s - loss: 0.7063 - add_all_ones: 111.7857
inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=220.0>

inside result... <tf.Variable 'total:0' shape=() dtype=float32, numpy=238.0>
16/16 [==============================] - 0s 17ms/step - loss: 0.7070 - add_all_ones: 133.0000
<tensorflow.python.keras.callbacks.History at 0x1d8213676c8>
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.4.1](https://colab.research.google.com/gist/amahendrakar/135940fbd8671cd3a6b8886bacdfa754/47099.ipynb) and TF-nightly. 

However, code works as expected with [TF v2.3.2](https://colab.research.google.com/gist/amahendrakar/dd908612f7e1cd7ba018feaa126ffe8a/47099-2-3.ipynb). Please check the linked gist for reference. Thanks!

**Comment by evgeniya-egupova** [comment]
Hi all! Looks like an issue described in #46713. Still awaiting response.

**Comment by ricvo** [comment]
Could that be a running average with some exp decay parameter? i.e. not a simple average.
Might be linked to https://github.com/tensorflow/tensorflow/issues/39448

**Comment by rchao** [comment]
Investigation into this shows that this is because Progbar doesn't have the correct stateful metrics information. Will work on a fix for this. Thanks for your patience.

**Comment by ricvo** [comment]
@rchao thanks for looking into this. Could you please tell in your opinion what the Progbar displayed metric values are at the moment?

**Comment by rchao** [comment]
@ricvo it would basically average the metric across batches, which is incorrect for stateful metrics like "AddAllOnes" used in the example repro.
